ext/socket/raddrinfo.c

```diff
@@ -463,37 +490,63 @@ rb_getaddrinfo_a(const char *node, const char *service,
     if (ret == 0)
         allocated_by_malloc = 1;
     else {
-	struct gai_suspend_arg arg;
-	struct gaicb *reqs[1];
-	struct gaicb req;
-
-	req.ar_name = node;
-	req.ar_service = service;
-	req.ar_request = hints;
-
-	reqs[0] = &req;
-	ret = getaddrinfo_a(GAI_NOWAIT, reqs, 1, NULL);
-	if (ret) return ret;
-        gaicbs_add(&req);
-
-	arg.req = &req;
-	arg.timeout = timeout;
-
-	ret = (int)(VALUE)rb_thread_call_without_gvl(nogvl_gai_suspend, &arg, RUBY_UBF_IO, 0);
-        gaicbs_remove(&req);
-        if (ret && ret != EAI_ALLDONE) {
-            /* EAI_ALLDONE indicates that the request already completed and gai_suspend was redundant */
-            /* on Ubuntu 18.04 (or other systems), gai_suspend(3) returns EAI_SYSTEM/ENOENT on timeout */
-            if (ret == EAI_SYSTEM && errno == ENOENT) {
-                return EAI_AGAIN;
-            } else {
-                return ret;
+        VALUE scheduler = rb_scheduler_current();
+
+        if (scheduler != Qnil && rb_scheduler_supports_address_resolve(scheduler)) {
+            VALUE host = rb_str_new_cstr(node);
+            VALUE rb_timeout = rb_time_timespec_new(timeout);
+            VALUE ip_addresses_array = rb_scheduler_address_resolve(scheduler, host, rb_timeout);
```

Comment:
This is `rb_getaddrinfo_a` function, the equivalent of `rb_getaddrinfo`. The only difference is it receives a `timeout` argument.
The rest of the logic here is exactly the same.